Enable configuration via yaml file + mavenExecute as example

kurzy opened the pull request.

o-liver commented on test/groovy/AbstractPiperUnitTest.groovy:
```diff
@@ -0,0 +1,28 @@
+import com.lesfurets.jenkins.unit.BasePipelineTest
+import org.junit.Before
+import org.yaml.snakeyaml.Yaml
```
Could not be resolved for me. Use instead:
`org.jenkinsci.plugins.pipeline.utility.steps.shaded.org.yaml.snakeyaml.Yaml`

o-liver replied:
This was solved by commit a92355c: "Update pom.xml"

o-liver commented on test/groovy/MavenExecuteTest.groovy:
```diff
@@ -0,0 +1,41 @@
+        def script = loadScript("test/resources/pipelines/mavenExecuteTest/executeBasicMavenCommand.groovy",)
+            script.execute()
+            TestCase.assertEquals('maven:3.5-jdk-7', dockerParameters.dockerImage)
+            TestCase.assertTrue(shellCalls.contains('mvn clean install'))
```
Test fails for me when running `mvn clean test`. Error message:
> `MavenExecuteTest.testExecuteBasicMavenCommand:29 expected:<maven:3.5-jdk-7> but was:<null>`

As a Unit Test it works, however.

o-liver commented on test/groovy/MavenExecuteTest.groovy:
```diff
@@ -0,0 +1,41 @@
+
+    @Test
+    void testExecuteBasicMavenCommand() throws Exception {
+        def script = loadScript("test/resources/pipelines/mavenExecuteTest/executeBasicMavenCommand.groovy",)
```
Why is there a comma after the string argument? (This is seen in all `loadScript` calls.)
This is why the following lines are indented, right?

marcusholl commented on src/com/sap/piper/DefaultValueCache.groovy:
```diff
@@ -0,0 +1,27 @@
+        return instance
+    }
+
+    static createInstance(Map defaultValues){
```
We are in a strong singe threaded world, I suppose? Otherwise we have to synchronize the method.

OliverNocon replied:
I guess this is OK. We should be in a single thread per job execution ...

marcusholl commented on test/groovy/AbstractPiperUnitTest.groovy:
```diff
@@ -0,0 +1,31 @@
+
+import static com.lesfurets.jenkins.unit.global.lib.LibraryConfiguration.library
+
+class AbstractPiperUnitTest extends BasePipelineTest {
```
There is already a base class merged in the meantime: test/groovy/PiperTestBase

kurzy replied:
I changed it to use PiperTestBase

marcusholl commented on vars/mavenExecute.groovy:
```diff
@@ -0,0 +1,86 @@
+                downloadSettingsFromUrl(globalSettingsFile)
+                globalSettingsFile = "settings.xml"
+            }
+            command += " --global-settings ${globalSettingsFile}"
```
Applies for all file pathes used in the command string: Does this also work when the file path contains blanks? Maybe we should quote the pathes.

kurzy replied:
It is now escaped

marcusholl commented:
For declarative pipeline it is - to my knowledge - not possible to use return values for steps. An approach to deal with that would be to update the configuration from step so that a later step an retrieve parameters from an earlier step (... or can be called with some values retrieved from the configuration, what is basically the same). Up to now there is no possibility to update the configuration. There are also reasons for having an immutable configuration. But maybe we can discuss how to deal with updates to the configuation from a step - maybe outside of that PR.

kurzy commented on resources/default_pipeline_environment.yml:
```diff
@@ -0,0 +1,8 @@
+#Steps Specific Configuration
+steps:
+  mavenExecute:
+    dockerImage: 'maven:3.5-jdk-7'
```
If you want to use a different registry, you just put the name in front of the docker image. That here takes it from the official docker hub. In most of our steps we want to use a default docker image.

kurzy commented on src/com/sap/piper/ConfigurationHelper.groovy:
```diff
@@ -0,0 +1,60 @@
+        this.config = config
+    }
+
+    def getConfigProperty(property) {
```
It is the same naming as it is already in commonPipelineEnvironment.
However, I rename it to key instead.

OliverNocon commented on vars/setupCommonPipelineEnvironment.groovy:
```diff
@@ -2,13 +2,38 @@ def call(Map parameters = [:]) {
+private loadConfigurationFromFile(script, String configFile) {
+
+    String defaultPropertiesYmlConfigFile = '.pipeline/config.properties'
+    String defaultYmlConfigFile = 'pipeline_config.yml'
```
can we move the default Config file also to .pipeline folder?

kurzy replied:
I am still in favor of having it in the root as the manifest.yml usually is also in the root and has a similar meaning, same with package.json and all the other configuration files for the tools.
Furthermore I see the problem that folders with dots in the beginning can be inconvenient for the user. For example:
* I personally have the feeling that a folder with dot does not belong to my project and should not be committed to git.
* On Windows in the explorer I cannot create a folder with a dot.
* On Linux/Mac these folders are treated as hidden folders. However, we see the configuration as essential part of the project and thus should be visible.

fwilhe replied:
What else does the `.pipeline` directory contain?

I think it is fair to compare the pipeline configuration to a `Makefile` or a `pom.xml`. Those files are not dotfiles, because they are a part of the software delivery. They describe how the software is built. The same holds true for the pipeline configuration, in my opinion.

I think about it this way: If I had a "export this source code repository to a zip file", what would I expect this repository to contain? Maybe not the `.gitigonore` file, because that is just an auxiliary file for git usage. Also maybe not a `.editorconfig` file, because this is just an auxiliary file. But surely, `Makefile`, `pom.xml`, `Jenkinsfile` and `pipeline_config.yml` needs to be in there, because they describe how the project is built, they *are* (part of) the software delivery.

I'm curious to know the rationale for `.pipeline`. If it only contains one file, the question is why this needs to be a directory?

kbarnold replied:
> I personally have the feeling that a folder with dot does not belong to my project and should not be committed to git.

IMO, we should not base our decisions on personal feelings. Files that should not be committed to Git are to be mentioned in a `.gitignore` file. There is no convention on the file name for this.

> On Windows in the explorer I cannot create a folder with a dot.

This is not true. Try creating a folder named `.pipelines.`. Windows will remove the trailing dot and you will have the expected outcome. I'm not saying that this is a good way of handling folder names starting with a dot, but it is an issue with Windows Explorer and should be fixed there.

> If I had a "export this source code repository to a zip file", what would I expect this repository to contain?

I would definitely not expect it to contain any configurations of deployment hosts. URLs of test systems could be strictly internal. In some companies leaking internal URLs is considered to be a security issue and must not happen.

> Maybe not the .gitigonore file, because that is just an auxiliary file for git usage

So why is `.gitignore` an auxiliary file for git usage but `Jenkinsfile` and `pipeline_config.yml` are not an auxiliary files for Jenkins usage? It's not like there are no alternatives to Jenkins. There are people who orchestrate using e.g. concourse.ci. One can switch that layer at all times and still build the same software, just like if they switched from Git to Mercurial.

marcusholl replied:
> I'm curious to know the rationale for .pipeline. If it only contains one file, the question is why this needs to be a directory?

Here I agree. In case we do not expect other files a .pipeline file would be sufficient.

Regarding the general question about dot files:
Some prominent examples on linux like systems are:
* ```.bash_profile``` contains settings for the bash
* ```.m2``` contains maven settings file, credentials with encrypted passwds and local repo 
* ```.mozilla``` Settings and cache for the firefox
* ```.kde```, a lot of settings etc. for KDE (in case KDE is used)
* ```.npm``` npm configurations
* ```.npmrc``` a cache
* ```.netrc``` credentials (can be used e.g. with curl)
* ```.ssh``` ssh settings, known hosts, private keys
* ```.gitignore``` (was already mentiones)

Now we should ask ourselfs: does ```.pipeline``` fits into that pattern?

Regarding ```pom.xml ```(controlling the build) is not a dot file: It is not a dot file, but ```.m2``` (containing among other things some fundamental maven settings (more the less build server specific)) is a dot file.

That relates IMO well to ```Jenkinsfile``` vs ```.pipeline```.

I personally think the semantics of our ```.pipeline``` file fits into the dotfile examples provided above.

OliverNocon replied:
There is also a reasoning for the folder ...

Teams can store additional files in there. We for example have team-specific pipeline extensions in there. They may also add other custom files there, additional config etc.
And that is what teams do today ...
Why should we change it for them?

kurzy replied:
Okay. I think one reason for the discussion is that in our projects the configuration file is used in a different way. Since, it is still configurable over the parameters, I suggest that we keep the folder .pipeline. I already changed the step loading it.

OliverNocon commented on src/com/sap/piper/DefaultValueCache.groovy:
```diff
@@ -0,0 +1,27 @@
+class DefaultValueCache implements Serializable {
+    private static DefaultValueCache instance
+
+    private Map defaultValues
```
I guess as Daniel put it, it is more Groovy like and should be ok

kurzy replied:
We could discuss whether we should omit Map here. I would not use generics as you try to minimize using types in groovy. Furthermore, I would not be really helpful, as it would be <String, Object> I guess.

marcusholl replied:
I would like to explain the reason for my question above. It is related to getDefaultValues below which returns the map.

The key question is: would we like to be sure that the default values are really what we configured accordingly in the persistency layer? If we say: "well, if somebody in a pipeline deliberately changes the default values from the persistency, it is OK for us ..." we can keep everything like it is. But if we say: "The default values must remain like they are configured in the persistency layer" then we have an issue with the current code. Again: I'm fine with the first statement, but we should be very clear about what we want.

When handing over the map to outside the caller of getDefaultValues can change the map, full CRUD. In case we would like to be sure about what is defined in persistency we can
1.) create a defensive copy. With that the caller cannot modify the defaults in the class, of course he can still modify his copy. Of course it needs to be a deep copy: a value in the defensive copy != corresponding value in the original (only important if the values are mutable).
2.) Create a unmodifiable map. With this the caller of the method can only read. See java.utils.Collections. But here we have the link to the generics here: This only makes sense if the values itself are immutable. Otherwise a caller cannot change the map, but the values contained in the map. with e.g. string we would be immutable. Of course it would be possible to create the unmodifiable map already in the constructor.

Having said all this I repeat what I already commented in the beginning: In case the approach with the default values does not mean that the defaults cannot be modified in the pipeline we can keep everything like it is.

kurzy replied:
I think, it is not a problem if someone modifies the default values in the pipeline. The Idea here was just to maintain these default values at a central place and not in the step.

kurzy commented:
@o-liver: I addressed your comments and added documentation for the new steps and some helper scripts.
@marcusholl: I addressed your comments or replied. I totally agree that PRs should be as small as possible and should only contain one new feature. However, since the PRs is already reviewed and created, I would not like to split it to reduce extra work, if you are fine with that.

marcusholl commented:
I was too much focused on the blanks - which can be solved by quoting.
I am aware of the differences of double quotes and singe quotes in bash.

For other characters I agree we need some escape helpers.

Again regarding the quotes: I would not mix up putting surrounding quotes around a variable and escaping some characters of a variable. I would like to suggest to only escape, but not to surround. With this approach we are more flexible and can also escape without getting something around. E.g. there might be use cases where someone would like to use double quotes in a shell script.

kurzy commented:
@marcusholl I am still in favor of using surrounding quotes as it solves the escaping for all special characters and we don't have to deal with them individually. As this should work for now I would suggest to leave it as it is and create an issue to change/improve the escape function when needed.

marcusholl commented:
I disagree. E.g. apache-commons StringEscapeUtils does not add surrounding quotes:

```
public class MyLauncher {

	public final static void main(String[] args) {

		String str = "Hello ' <title>SAP</title> & World!";
		System.out.println(org.apache.commons.lang3.StringEscapeUtils.ESCAPE_XML11.translate(str));
		System.out.println(org.apache.commons.lang3.StringEscapeUtils.ESCAPE_CSV.translate(str));
		System.out.println(org.apache.commons.lang3.StringEscapeUtils.ESCAPE_JAVA.translate(str));
		System.out.println(org.apache.commons.lang3.StringEscapeUtils.ESCAPE_JSON.translate(str));
		System.out.println(org.apache.commons.lang3.StringEscapeUtils.ESCAPE_ECMASCRIPT.translate(str));
	}
}
```

returns
```
Hello &apos; &lt;title&gt;SAP&lt;/title&gt; &amp; World!
Hello ' <title>SAP</title> & World!
Hello ' <title>SAP</title> & World!
Hello ' <title>SAP<\/title> & World!
Hello \' <title>SAP<\/title> & World!
```
There are no surrounding quotes.

Having it without quotes fits also best into the [principle of least of least surprise](https://en.wikipedia.org/wiki/Principle_of_least_astonishment)

Without surrounding quotes it leads also to better readable shell statements. Having the quotes (single, double (depending on the use case)) increases the situation awareness of what really happens in the shell.

>I am still in favor of using surrounding quotes as it solves the escaping for all special characters

Yes, with having single quotes the remaining char to escape is the single quotes (as far as I can see). But as stated above: it is better to see that single quote in the shell statement explicitly IMO.

In order to avoid a two-person-battle I would like to get opinions also from other colleagues.

OliverNocon commented:
Hi all,

I now also had a closer look.
I agree with Markus to have the single quotes in the call directly instead of in the escape function.

This makes it more transparent in my opinion.
Since this is not related to the config topic we could merge without this change if you don't find time to work on it and bring it in later. What do you think?

Best regards,
Oliver

kurzy commented:
@kbarnold Thanks. Documentation is fixed.
@OliverNocon @marcusholl I removed the escape function for now and added single quotes in the maven command directly. I suggest we create a new escape helper function in another pull request to solve escaping special characters in passwords and so on.

OliverNocon commented:
Thanks!
